### PR TITLE
CMake: Looking for THREADS when building tests

### PR DESCRIPTION
**Pull request by thopiekar**
Compared to the 4.0 branch, there is the instruction to link against pthread now.
However, pthreads is (as it is named) only available on posix (except you are using MinGW on Windows).
Anyway, I disabled hereby the usage of (p)threads on Windows. Probably there is a more elegant way to handle this situation, but that will allow Windows builds again at least.

### CODE REVIEW

**Comment by thopiekar** [comment]
@Ghostkeeper While doing some research I noticed that you added the pthread line before.
Since you are more experienced and know why the change was needed here, I guess you'll have an opinion about my work here 😅

**Comment by Ghostkeeper** [comment]
I added that because I was getting linking errors with pthread when compiling the GoogleTest test cases.

We're not using MinGW for this plug-in since it needs to link with CPython. It needs to be compiled with MSVC.

So I'd like to suggest a change to this then: Don't make `BUILD_WITH_THREADS` an option, but link with pthreads only if building tests (`if(BUILD_TESTS)`). I should've done that in the first place. I believe `BUILD_TESTS` is off by default? Otherwise you could disable it by default on Windows.

**Comment by thopiekar** [comment]
Ok. Before I begin: Where is the pthreads requirement coming from? Is it really been used within the tests directly or is it needed since GTest relies on it?

Yes, it is off by default, but I'm still working on my forked build environment and therefore I noticed that issue here.

File: CMakeLists.txt [hunk 1]
```diff
@@ -7,6 +7,16 @@ include(CMakePackageConfigHelpers)
...
 
+if(WIN32)
+  set(BUILD_WITH_THREADS_DEFAULT OFF)
+else()
+  set(BUILD_WITH_THREADS_DEFAULT ON)
+endif()
+option(BUILD_WITH_THREADS "Whether to link against threads or not." ${BUILD_WITH_THREADS_DEFAULT})
+
+if(BUILD_WITH_THREADS)
+    find_package(Threads)
```
Comment (LipuFei) [inline CMakeLists.txt]: I think it's more robust to do `find_package(Threads REQUIRED)` here. If it's not found, the line lower `target_link_libraries(Savitar PUBLIC pthread pugixml)` will fail.

File: CMakeLists.txt [hunk 2]
```diff
@@ -67,13 +76,16 @@ else()
     add_library(Savitar SHARED ${savitar_SRCS})
 endif()
 
-target_link_libraries(Savitar PUBLIC pthread pugixml)
+if(BUILD_WITH_THREADS)
+    target_link_libraries(Savitar PUBLIC pthread pugixml)
```
Comment (LipuFei) [inline CMakeLists.txt]: You can use `CMAKE_THREAD_LIBS_INIT`. Example: `target_link_libraries(Savitar PUBLIC CMAKE_THREAD_LIBS_INIT pugixml)`. See https://cmake.org/cmake/help/v3.10/module/FindThreads.html

File: CMakeLists.txt (hunk 1 above)
Comment (thopiekar) [reply CMakeLists.txt]: Makes sense 👍

**Comment by Ghostkeeper** [comment]
> Where is the pthreads requirement coming from? Is it really been used within the tests directly or is it needed since GTest relies on it?

Good question, but GTest has to be included with the headers of the main code as well since you need to be able to define friend classes sometimes. Maybe that's it.

**Comment by thopiekar** [comment]
While looking into https://cmake.org/cmake/help/v3.0/module/FindGTest.html I can imagine that it might be included in one of their variables.
So maybe depending on how GTest was built `pthread` might appear in their variables.
However, if I can't find anything I will go the way as @LipuFei and you suggested first.

I'm busy for some time but think I'll be able to make the corrections soon this week. 😉

**Comment by thopiekar** [comment]
Seems to be fine for me now. When building with tests on Windows I get no complaints about pthread.
However, there is one of three tests which need fixes:
```
ThreeMFParserTest.cpp
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(57): error C2079: "expected_verts" verwendet undefiniertes class "std::array<size_t,4>"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(57): error C2440: "Initialisierung": "initializer list" kann nicht in "int" konvertiert werden
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(57): note: Der Initialisierer enthält zu viele Elemente.
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(58): error C2079: "expected_tris" verwendet undefiniertes class "std::array<size_t,4>"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(58): error C2440: "Initialisierung": "initializer list" kann nicht in "int" konvertiert werden
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(58): note: Der Initialisierer enthält zu viele Elemente.
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(59): error C2079: "expected_child" verwendet undefiniertes class "std::array<size_t,4>"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(59): error C2440: "Initialisierung": "initializer list" kann nicht in "int" konvertiert werden
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(59): note: Der Initialisierer enthält zu viele Elemente.
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(67): error C2109: Index erfordert ein Array oder einen Zeigertyp
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(67): error C2660: "testing::internal::EqHelper<false>::Compare": Funktion akzeptiert keine 3 Argumente
D:\User\GIT\CDK\install\include\gtest/gtest.h(1481): note: Siehe Deklaration von "testing::internal::EqHelper<false>::Compare"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(67): error C2512: "testing::AssertionResult": Kein geeigneter Standardkonstruktor verfügbar
D:\User\GIT\CDK\install\include\gtest/gtest.h(281): note: Siehe Deklaration von "testing::AssertionResult"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(70): error C2109: Index erfordert ein Array oder einen Zeigertyp
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(70): error C2660: "testing::internal::EqHelper<false>::Compare": Funktion akzeptiert keine 3 Argumente
D:\User\GIT\CDK\install\include\gtest/gtest.h(1481): note: Siehe Deklaration von "testing::internal::EqHelper<false>::Compare"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(70): error C2512: "testing::AssertionResult": Kein geeigneter Standardkonstruktor verfügbar
D:\User\GIT\CDK\install\include\gtest/gtest.h(281): note: Siehe Deklaration von "testing::AssertionResult"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(73): error C2109: Index erfordert ein Array oder einen Zeigertyp
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(73): error C2660: "testing::internal::EqHelper<false>::Compare": Funktion akzeptiert keine 3 Argumente
D:\User\GIT\CDK\install\include\gtest/gtest.h(1481): note: Siehe Deklaration von "testing::internal::EqHelper<false>::Compare"
D:\User\GIT\CDK\build\libSavitar-prefix\src\libSavitar\tests\ThreeMFParserTest.cpp(73): error C2512: "testing::AssertionResult": Kein geeigneter Standardkonstruktor verfügbar
D:\User\GIT\CDK\install\include\gtest/gtest.h(281): note: Siehe Deklaration von "testing::AssertionResult"
```

But yeah, that is not my fault 😋

**Comment by thopiekar** [comment]
Sure. No problem 😉

**Pull request merged**